tslearn/nonmyopic.py

```diff
@@ -0,0 +1,370 @@
+         float : the so fat minimum cost
+        """
+        tau_star = 0
+        so_far_minimum_cost = function_to_minimize(xt, tau_star)
+        for tau in range(1, end_of_time):
+            current_cost = function_to_minimize(xt, tau)
+            if current_cost < so_far_minimum_cost:
+                so_far_minimum_cost = current_cost
+                tau_star = tau
+        return tau_star + xt.shape[-1], so_far_minimum_cost
+
+    def _predict_at_fixed_length(self, xt, cost_function):
+        """
+        This function classifies a single time series xt
+
+        Parameters
+        ----------
+        xt: vector
+            a time series that is probably incomplete but that nonetheless we want to classify
+        cost_function: funct
+            The cost function that will be use to achieve our classification purpose
+        Returns
+        -------
+        int: the class which is predicted
+        int : the time of the prediction
+        float : the probability used for computing the cost
+        float : the loss when classifying
+        """
+        time_stamp = xt.shape[0]
+        xt = np.reshape(xt, (1, time_stamp))
+        time_prediction = self.minimum_time_stamp
+        stop_criterion = False
+        while stop_criterion is False:
+            minimum_tau, minimum_loss = self.minimize_integer(
+                end_of_time=time_stamp - time_prediction,
+                function_to_minimize=cost_function,
+                xt=xt[:, :time_prediction],
+            )
+            if time_prediction == time_stamp:
+                result = self.classifier_[-1].predict(xt)
+                result_proba = self.classifier_[-1].predict_proba(xt)
+                stop_criterion = True
+                minimum_tau, minimum_loss = self.minimize_integer(
+                    end_of_time=time_stamp - time_prediction,
+                    function_to_minimize=cost_function,
+                    xt=xt[:, :time_prediction],
+                )
+                loss_exit = minimum_loss
+            elif minimum_tau == time_prediction:
+                result = self.classifier_[
+                    time_prediction - self.minimum_time_stamp
+                ].predict(xt[:, :time_prediction])
+                result_proba = self.classifier_[
+                    time_prediction - self.minimum_time_stamp
+                ].predict_proba(xt[:, :time_prediction])
+                loss_exit = minimum_loss
+                stop_criterion = True
+            else:
+                time_prediction = time_prediction + 1
+        return result, time_prediction, result_proba, loss_exit
+
+    def predict(self, X):
+        """
+        Predicts the classes of the series of the test dataset.
+
+        Parameters
+        ----------
+        X_test : Array-like
+            The test dataset
+
+        Returns
+        -------
+        Vector : the predicted classes
+        """
+
+        Y_pred = []
```

Comment:
The `sklearn` standard is to write `y` instead of `Y` for label arrays. Please change those in your code (here and in other places too).